106-auto-rebalance-on-imbalanced-clusters.md

```diff
@@ -0,0 +1,471 @@
+With this proposal, we are only going to support auto-rebalance on imbalanced cluster.
+We also plan to implement the same for topic and metrics related issues, but it will be part of future work since their implementation require different approach.
+For example, when dealing with topic related issues, it will require a coordination with topic operator and metrics issues will require coordination with the Kafka apis.
+
+Here is an example of what the configured `Kafka` custom resource could look like:
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+  kafka:
+  # ...
+  cruiseControl:
+    # ...
+    autoRebalance:
+      # using the skew rebalance mode
+      - mode: skew
+        template:
+          name: my-skew-rebalance-template
+```
+
+It is also possible to use the default Cruise Control rebalancing configuration by omitting the `template` field.
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+  kafka:
+    # ...
+  cruiseControl:
+    # ...
+    autoRebalance:
+      # using the default Cruise Control rebalancing configuration 
+      - mode: skew
+```
+
+The auto-rebalance configuration for the `spec.cruiseControl.autoRebalance.template` property in the `Kafka` custom resource is provided through a `KafkaRebalance` custom resource defined as a "template".
+That is a `KafkaRebalance` custom resource with the `strimzi.io/rebalance-template: true` annotation set.
+When it is created, the `KafkaRebalanceAssemblyOperator` doesn't run any rebalancing.
+This is because it doesn't represent an "actual" rebalance request to get an optimization proposal, but it's just the place where configuration related to auto-rebalancing is defined.
+The user can specify rebalancing goals and other configuration for rebalancing, within the resource.
+
+Here is an example template:
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaRebalance
+metadata:
+  name: my-skew-rebalance-template
+  annotations:
+    strimzi.io/rebalance-template: "true" # specifies that this KafkaRebalance is a rebalance configuration template
+spec:
+  goals:
+    - CpuCapacityGoal
+    - NetworkInboundCapacityGoal
+    - DiskCapacityGoal
+    - RackAwareGoal
+    - MinTopicLeadersPerBrokerGoal
+    - NetworkOutboundCapacityGoal
+    - ReplicaCapacityGoal
+  skipHardGoalCheck: true
+  # ... other rebalancing related configuration
+```
+When the "template" is set the operator automatically creates (or updates) a corresponding "actual" `KafkaRebalance` custom resource based on the "template" when an anomaly is detected and notified by the `AnomalyDetectorNotifier`
+The operator copies over goals and rebalancing options from the referenced "template" resource to the "actual" rebalancing one and also adds the `spec.mode` to it.
```

Comment:
"adds the `spec.mode` to it." ... with auto-rebalancing it was just a copy of the mode field.
Now it's not a copy, because "skew" doesn't mean anything to a `KafkaRebalance` custom resource but it has to be "full". I would mention that the "skew" mode will be mapped to the "full" mode into the actual `KafkaRebalance` resource.